File: sdk/java/src/main/java/dev/rue/hook/Hooks.java
package dev.rue.hook;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * An interface per kind, and the dispatch from a request frame to a handler.
 *
 * <p>The registration frame is built from the kinds actually supplied, so a
 * hook cannot register for one it does not serve -- the failure that produces
 * is a plan binding to it and refusing at its first step, a long way from
 * where the mistake was made.
 *
 * <p>A handler returns its answer or throws {@link Refusal}. A refusal is a
 * hook saying no, on the record, and is not a fault; any other exception
 * becomes one too, because an unanswered request is Silent and Silent tells
 * the operator nothing about why.
 */
public final class Hooks {

    /** Why a hook will not answer: the text the engine journals. */
    public static class Refusal extends RuntimeException {
        public Refusal(String why) {
            super(why);
        }

        public static Refusal unserved(String kind, String op) {
            return new Refusal("this hook does not serve " + kind + "." + op);
        }
    }

    public interface Journal {
        void append(Object entry);
    }

    public interface Inventory {
        List<Object> list();
    }

    /** The instance-directory ops below {@code run} are required only of a
     * hook that registers {@code filesystem} (7.7). */
    public interface Execute {
        Map<String, Object> run(String host, String instance, List<Object> body);

        /** The file's text, or null for no such file -- an answer, not a refusal. */
        default String readFact(String host, String shape) {
            throw Refusal.unserved("execute", "read_fact");
        }

        default Map<String, Object> bootstrapState(String host) {
            throw Refusal.unserved("execute", "bootstrap_state");
        }

        /** The one op the engine reads a refusal of as "no skew probe is
         * possible here" rather than as a fault. */
        default long clock(String host) {
            throw Refusal.unserved("execute", "clock");
        }

        default void instanceDirCreate(String host, String instance) {
            throw Refusal.unserved("execute", "instance_dir_create");
        }

        default void instanceDirRemove(String host, String instance) {
            throw Refusal.unserved("execute", "instance_dir_remove");
        }

        default List<Object> instanceDirList(String host) {
            throw Refusal.unserved("execute", "instance_dir_list");
        }

        default void putFile(String host, String instance, String rel, String content, long mode) {
            throw Refusal.unserved("execute", "put_file");
        }

        default void replaceFile(String host, String instance, String rel, String content) {
            throw Refusal.unserved("execute", "replace_file");
        }

        default String getFile(String host, String instance, String rel) {
            throw Refusal.unserved("execute", "get_file");
        }

        default void removeFile(String host, String instance, String rel) {
            throw Refusal.unserved("execute", "remove_file");
        }

        default void hostLock(String host) {
            throw Refusal.unserved("execute", "host_lock");
        }
    }

    public interface Probe {
        /** Use {@link #observation}. */
        Map<String, Object> observe(String host, String probe);
    }

    public interface Approval {
        List<Object> authenticators();

        String challenge(String instance, String digest, Object scope, Object context);

        /** {@code {verified, reason}}; use {@link #verdict}. */
        Map<String, Object> verify(
                String instance, String digest, Object scope, String authenticator, String proof);
    }

    public interface Secrets {
        default String resolve(String reference) {
            throw Refusal.unserved("secrets", "resolve");
        }

        /** {@code {accepted, receipt}}; use {@link #delivery}. */
        default Map<String, Object> deliver(String instance, String label, String value) {
            throw Refusal.unserved("secrets", "deliver");
        }
    }

    public interface Notify {
        void deliver(String level, String subject, String body);
    }

    public interface Scheduler {
        void install(String host, String artifact);

        void arm(String host, String artifact, Object deadline);

        void rearm(String host, String artifact, Object deadline);

        void disarm(String host, String artifact);

        /** true, false, or the string "unknown". Never guess: the engine
         * reads false as "install it again". */
        Object present(String host, String artifact);
    }

    // --- little constructors, so a handler never hand-builds a wire map ---

    public static Map<String, Object> observation(String tri, String text) {
        return Map.of("text", text, "tri", tri);
    }

    public static Map<String, Object> verdict(boolean verified, String reason) {
        return Map.of("verified", verified, "reason", reason);
    }

    public static Map<String, Object> delivery(boolean accepted, String receipt) {
        return Map.of("accepted", accepted, "receipt", receipt);
    }

    public static Map<String, Object> authenticator(String id, boolean human) {
        return Map.of("id", id, "human", human);
    }

    public static Map<String, Object> output(String stdout, Map<String, Object> outputs) {
        return Map.of("stdout", stdout, "outputs", outputs);
    }

    /** The text of a resolved value. Named rather than reached through the
     * map so that reading a secret is a visible act in the code that does it. */
    @SuppressWarnings("unchecked")
    public static String expose(Object resolved) {
        if (resolved instanceof Map<?, ?> m) {
            return String.valueOf(((Map<String, Object>) m).get("text"));
        }
        return String.valueOf(resolved);
    }

    // --- the hooks themselves --------------------------------------------

    public Journal journal;
    public Inventory inventory;
    public Execute execute;
    public Probe probe;
    public Approval approval;
    public Secrets secrets;
    public Notify notify;
    public Scheduler scheduler;
    /** This hook serves the instance-directory ops (7.7). */
    public boolean filesystem;
    public boolean stdinPreamble;

    public List<String> kinds() {
        List<String> k = new ArrayList<>();
        if (journal != null) k.add("journal");
        if (inventory != null) k.add("inventory");
        if (execute != null) k.add("execute");
        if (probe != null) k.add("probe");
        if (approval != null) k.add("approval");
        if (secrets != null) k.add("secrets");
        if (notify != null) k.add("notify");
        if (scheduler != null) k.add("scheduler");
        return k;
    }

    /**
     * Answer one request frame. The reply always carries the request's id,
     * including for an op with no handler: a request left unanswered is
     * Silent, and the engine can only report that as "the step did not
     * happen".
     */
    public Map<String, Object> answer(Map<String, Object> request) {
        Object id = request.get("id");
        String kind = str(request, "kind");
        String opName = str(request, "op");
        Op row = Op.find(kind, opName);
        if (row == null) {
            return refusal(id, kind + "." + opName + " is not an op of this protocol");
        }
        Map<String, Object> fields;
        try {
            fields = dispatch(row, request);
        } catch (Refusal why) {
            return refusal(id, why.getMessage());
        } catch (RuntimeException e) {
            return refusal(id, e.getClass().getSimpleName() + ": " + e.getMessage());
        }
        Map<String, Object> reply = new LinkedHashMap<>();
        reply.put("id", id);
        reply.put("ok", true);
        reply.putAll(fields);
        for (String f : row.requiredReply()) {
            if (!reply.containsKey(f)) {
                // Replies are built from the op's row, so this is a bug here
                // and not an R0303 for the far end to puzzle over.
                return refusal(id, "the SDK built a " + kind + "." + opName + " reply without " + f);
            }
        }
        return reply;
    }

    private static Map<String, Object> refusal(Object id, String why) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("id", id);
        m.put("ok", false);
        m.put("error", why);
        return m;
    }

    private static String str(Map<String, Object> m, String k) {
        Object v = m.get(k);
        return v == null ? "" : String.valueOf(v);
    }

    private static Map<String, Object> one(String k, Object v) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put(k, v);
        return m;
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> dispatch(Op row, Map<String, Object> r) {
        String host = str(r, "host");
        String inst = str(r, "instance");
        switch (row.kind()) {
            case "journal" -> {
                if (journal == null) throw Refusal.unserved(row.kind(), row.op());
                Object entry = r.get("entry");
                if (entry == null) throw new Refusal("journal.append without an entry");
                journal.append(entry);
                return Map.of();
            }
            case "inventory" -> {
                if (inventory == null) throw Refusal.unserved(row.kind(), row.op());
                return one("hosts", inventory.list());
            }
            case "probe" -> {
                if (probe == null) throw Refusal.unserved(row.kind(), row.op());
                return one("fact", probe.observe(host, str(r, "probe")));
            }
            case "notify" -> {
                if (notify == null) throw Refusal.unserved(row.kind(), row.op());
                notify.deliver(str(r, "level"), str(r, "subject"), str(r, "body"));
                return Map.of();
            }
            case "approval" -> {
                if (approval == null) throw Refusal.unserved(row.kind(), row.op());
                return switch (row.op()) {
                    case "authenticators" -> one("authenticators", approval.authenticators());
                    case "challenge" -> one("challenge",
                            approval.challenge(inst, str(r, "digest"), r.get("scope"), r.get("context")));
                    default -> approval.verify(inst, str(r, "digest"), r.get("scope"),
                            str(r, "authenticator"), str(r, "proof"));
                };
            }
            case "secrets" -> {
                if (secrets == null) throw Refusal.unserved(row.kind(), row.op());
                if (row.op().equals("resolve")) {
                    return one("value", secrets.resolve(str(r, "ref")));
                }
                return secrets.deliver(inst, str(r, "label"), str(r, "value"));
            }
            case "scheduler" -> {
                if (scheduler == null) throw Refusal.unserved(row.kind(), row.op());
                Object deadline = r.get("deadline");
                String artifact = str(r, "artifact");
                switch (row.op()) {
                    case "install" -> scheduler.install(host, artifact);
                    case "arm" -> scheduler.arm(host, artifact, deadline);
                    case "rearm" -> scheduler.rearm(host, artifact, deadline);
                    case "disarm" -> scheduler.disarm(host, artifact);
                    default -> {
                        return one("present", scheduler.present(host, artifact));
                    }
                }
                return Map.of();
            }
            default -> {
                if (execute == null) throw Refusal.unserved(row.kind(), row.op());
                switch (row.op()) {
                    case "run" -> {
                        List<Object> body = (List<Object>) r.getOrDefault("body", List.of());
                        return one("output", execute.run(host, inst, body));
                    }
                    case "read_fact" -> {
                        String c = execute.readFact(host, str(r, "shape"));
                        // null is "no such file", which is an answer; an
                        // empty string would be a file that exists and is
                        // empty.
                        return c == null ? Map.of() : one("content", c);
                    }
                    case "bootstrap_state" -> {
                        return one("state", execute.bootstrapState(host));
                    }
                    case "clock" -> {
                        return one("epoch_s", execute.clock(host));
                    }
                    case "instance_dir_create" -> execute.instanceDirCreate(host, inst);
                    case "instance_dir_remove" -> execute.instanceDirRemove(host, inst);
                    case "instance_dir_list" -> {
                        return one("dirs", execute.instanceDirList(host));
                    }
                    case "put_file" -> execute.putFile(host, inst, str(r, "rel"), str(r, "content"),
                            r.get("mode") instanceof Long l ? l : 0L);
                    case "replace_file" -> execute.replaceFile(host, inst, str(r, "rel"), str(r, "content"));
                    case "get_file" -> {
                        return one("content", execute.getFile(host, inst, str(r, "rel")));
                    }
                    case "remove_file" -> execute.removeFile(host, inst, str(r, "rel"));
                    default -> execute.hostLock(host);
                }
                return Map.of();
            }
        }
    }
}

File: sdk/dotnet/src/RueHook/Hooks.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Rue.Hook;

/// <summary>Why a hook will not answer: the text the engine journals and
/// the operator reads. Throwing it is a hook saying no, on the record, and
/// is not a fault.</summary>
public sealed class Refusal : Exception
{
    public Refusal(string why) : base(why) { }

    public static Refusal Unserved(string kind, string op) =>
        new($"this hook does not serve {kind}.{op}");
}

public interface IJournal
{
    void Append(JsonNode entry);
}

public interface IInventory
{
    IEnumerable<object> List();
}

/// The instance-directory ops below <c>Run</c> are required only of a hook
/// that registers <c>filesystem</c> (7.7).
public interface IExecute
{
    object Run(string host, string instance, IReadOnlyList<JsonNode?> body);

    /// The file's text, or null for no such file -- an answer, not a refusal.
    string? ReadFact(string host, string shape) => throw Refusal.Unserved("execute", "read_fact");

    object BootstrapState(string host) => throw Refusal.Unserved("execute", "bootstrap_state");

    /// The one op the engine reads a refusal of as "no skew probe is
    /// possible here" rather than as a fault.
    long Clock(string host) => throw Refusal.Unserved("execute", "clock");

    void InstanceDirCreate(string host, string instance) =>
        throw Refusal.Unserved("execute", "instance_dir_create");

    void InstanceDirRemove(string host, string instance) =>
        throw Refusal.Unserved("execute", "instance_dir_remove");

    IEnumerable<object> InstanceDirList(string host) =>
        throw Refusal.Unserved("execute", "instance_dir_list");

    void PutFile(string host, string instance, string rel, string content, long mode) =>
        throw Refusal.Unserved("execute", "put_file");

    void ReplaceFile(string host, string instance, string rel, string content) =>
        throw Refusal.Unserved("execute", "replace_file");

    string GetFile(string host, string instance, string rel) =>
        throw Refusal.Unserved("execute", "get_file");

    void RemoveFile(string host, string instance, string rel) =>
        throw Refusal.Unserved("execute", "remove_file");

    void HostLock(string host) => throw Refusal.Unserved("execute", "host_lock");
}

public interface IProbe
{
    object Observe(string host, string probe);
}

public interface IApproval
{
    IEnumerable<object> Authenticators();

    string Challenge(string instance, string digest, JsonNode? scope, JsonNode? context);

    object Verify(string instance, string digest, JsonNode? scope, string authenticator, string proof);
}

public interface ISecrets
{
    string Resolve(string reference) => throw Refusal.Unserved("secrets", "resolve");

    object Deliver(string instance, string label, string value) =>
        throw Refusal.Unserved("secrets", "deliver");
}

public interface INotify
{
    void Deliver(string level, string subject, string body);
}

public interface IScheduler
{
    void Install(string host, string artifact);

    void Arm(string host, string artifact, JsonNode? deadline);

    void Rearm(string host, string artifact, JsonNode? deadline);

    void Disarm(string host, string artifact);

    /// true, false, or the string "unknown". Never guess: the engine reads
    /// false as "install it again", which is destructive when it is wrong.
    object Present(string host, string artifact);
}

/// <summary>
/// What this hook serves, and the dispatch from a request frame to a handler.
///
/// The registration frame is built from the kinds actually supplied, so a
/// hook cannot register for one it does not serve -- the failure that
/// produces is a plan binding to it and refusing at its first step, a long
/// way from where the mistake was made.
/// </summary>
public sealed class Hooks
{
    public IJournal? Journal { get; set; }
    public IInventory? Inventory { get; set; }
    public IExecute? Execute { get; set; }
    public IProbe? Probe { get; set; }
    public IApproval? Approval { get; set; }
    public ISecrets? Secrets { get; set; }
    public INotify? Notify { get; set; }
    public IScheduler? Scheduler { get; set; }

    /// This hook serves the instance-directory ops (7.7).
    public bool Filesystem { get; set; }

    public bool StdinPreamble { get; set; }

    public List<string> Kinds()
    {
        var k = new List<string>();
        if (Journal is not null) k.Add("journal");
        if (Inventory is not null) k.Add("inventory");
        if (Execute is not null) k.Add("execute");
        if (Probe is not null) k.Add("probe");
        if (Approval is not null) k.Add("approval");
        if (Secrets is not null) k.Add("secrets");
        if (Notify is not null) k.Add("notify");
        if (Scheduler is not null) k.Add("scheduler");
        return k;
    }

    // --- little constructors, so a handler never hand-builds a wire map ---

    public static object Observation(string tri, string text = "") =>
        new Dictionary<string, object?> { ["text"] = text, ["tri"] = tri };

    public static object Verdict(bool verified, string reason) =>
        new Dictionary<string, object?> { ["verified"] = verified, ["reason"] = reason };

    public static object Delivery(bool accepted, string receipt) =>
        new Dictionary<string, object?> { ["accepted"] = accepted, ["receipt"] = receipt };

    public static object Authenticator(string id, bool human) =>
        new Dictionary<string, object?> { ["id"] = id, ["human"] = human };

    public static object Output(string stdout, IDictionary<string, object?> outputs) =>
        new Dictionary<string, object?> { ["stdout"] = stdout, ["outputs"] = outputs };

    /// The text of a resolved value. Named rather than reached through the
    /// node so that reading a secret is a visible act in the code that does it.
    public static string Expose(JsonNode? resolved) =>
        resolved?["text"]?.GetValue<string>() ?? string.Empty;

    /// <summary>
    /// Answer one request frame. The reply always carries the request's id,
    /// including for an op with no handler: a request left unanswered is
    /// Silent, and the engine can only report that as "the step did not
    /// happen".
    /// </summary>
    public JsonObject Answer(JsonObject request)
    {
        var id = request["id"]?.DeepClone();
        var kind = Str(request, "kind");
        var opName = Str(request, "op");
        var row = Op.Find(kind, opName);
        if (row is null)
        {
            return RefusalFrame(id, $"{kind}.{opName} is not an op of this protocol");
        }

        Dictionary<string, object?> fields;
        try
        {
            fields = Dispatch(row, request);
        }
        catch (Refusal why)
        {
            return RefusalFrame(id, why.Message);
        }
        catch (Exception e)
        {
            return RefusalFrame(id, $"{e.GetType().Name}: {e.Message}");
        }

        var reply = new JsonObject { ["id"] = id, ["ok"] = true };
        foreach (var (k, v) in fields)
        {
            reply[k] = JsonSerializer.SerializeToNode(v);
        }

        foreach (var f in row.RequiredReply)
        {
            if (reply[f] is null)
            {
                // Replies are built from the op's row, so this is a bug here
                // and not an R0303 for the far end to puzzle over.
                return RefusalFrame(id, $"the SDK built a {kind}.{opName} reply without {f}");
            }
        }

        return reply;
    }

    private static JsonObject RefusalFrame(JsonNode? id, string why) =>
        new() { ["id"] = id, ["ok"] = false, ["error"] = why };

    private static string Str(JsonObject o, string k) =>
        o[k] is { } n && n.GetValueKind() == JsonValueKind.String ? n.GetValue<string>() : "";

    private static Dictionary<string, object?> One(string k, object? v) => new() { [k] = v };

    private static Dictionary<string, object?> None() => new();

    private Dictionary<string, object?> Dispatch(Op row, JsonObject r)
    {
        var host = Str(r, "host");
        var inst = Str(r, "instance");
        switch (row.Kind)
        {
            case "journal":
                {
                    var h = Journal ?? throw Refusal.Unserved(row.Kind, row.OpName);
                    var entry = r["entry"] ?? throw new Refusal("journal.append without an entry");
                    h.Append(entry);
                    return None();
                }
            case "inventory":
                {
                    var h = Inventory ?? throw Refusal.Unserved(row.Kind, row.OpName);
                    return One("hosts", h.List());
                }
            case "probe":
                {
                    var h = Probe ?? throw Refusal.Unserved(row.Kind, row.OpName);
                    return One("fact", h.Observe(host, Str(r, "probe")));
                }
            case "notify":
                {
                    var h = Notify ?? throw Refusal.Unserved(row.Kind, row.OpName);
                    h.Deliver(Str(r, "level"), Str(r, "subject"), Str(r, "body"));
                    return None();
                }
            case "approval":
                {
                    var h = Approval ?? throw Refusal.Unserved(row.Kind, row.OpName);
                    return row.OpName switch
                    {
                        "authenticators" => One("authenticators", h.Authenticators()),
                        "challenge" => One("challenge",
                            h.Challenge(inst, Str(r, "digest"), r["scope"], r["context"])),
                        _ => Flatten(h.Verify(inst, Str(r, "digest"), r["scope"],
                            Str(r, "authenticator"), Str(r, "proof")))
                    };
                }
            case "secrets":
                {
                    var h = Secrets ?? throw Refusal.Unserved(row.Kind, row.OpName);
                    return row.OpName == "resolve"
                        ? One("value", h.Resolve(Str(r, "ref")))
                        : Flatten(h.Deliver(inst, Str(r, "label"), Str(r, "value")));
                }
            case "scheduler":
                {
                    var h = Scheduler ?? throw Refusal.Unserved(row.Kind, row.OpName);
                    var artifact = Str(r, "artifact");
                    switch (row.OpName)
                    {
                        case "install": h.Install(host, artifact); return None();
                        case "arm": h.Arm(host, artifact, r["deadline"]); return None();
                        case "rearm": h.Rearm(host, artifact, r["deadline"]); return None();
                        case "disarm": h.Disarm(host, artifact); return None();
                        default: return One("present", h.Present(host, artifact));
                    }
                }
            default:
                {
                    var h = Execute ?? throw Refusal.Unserved(row.Kind, row.OpName);
                    switch (row.OpName)
                    {
                        case "run":
                            {
                                var body = r["body"] as JsonArray ?? new JsonArray();
                                return One("output", h.Run(host, inst, body.ToList()));
                            }
                        case "read_fact":
                            {
                                var c = h.ReadFact(host, Str(r, "shape"));
                                // null is "no such file", which is an answer;
                                // an empty string would be a file that exists
                                // and is empty.
                                return c is null ? None() : One("content", c);
                            }
                        case "bootstrap_state": return One("state", h.BootstrapState(host));
                        case "clock": return One("epoch_s", h.Clock(host));
                        case "instance_dir_create": h.InstanceDirCreate(host, inst); return None();
                        case "instance_dir_remove": h.InstanceDirRemove(host, inst); return None();
                        case "instance_dir_list": return One("dirs", h.InstanceDirList(host));
                        case "put_file":
                            h.PutFile(host, inst, Str(r, "rel"), Str(r, "content"),
                                r["mode"]?.GetValue<long>() ?? 0);
                            return None();
                        case "replace_file":
                            h.ReplaceFile(host, inst, Str(r, "rel"), Str(r, "content"));
                            return None();
                        case "get_file": return One("content", h.GetFile(host, inst, Str(r, "rel")));
                        case "remove_file": h.RemoveFile(host, inst, Str(r, "rel")); return None();
                        default: h.HostLock(host); return None();
                    }
                }
        }
    }

    /// A handler that returns a whole reply body (a verdict, a delivery)
    /// hands back its fields rather than one named value.
    private static Dictionary<string, object?> Flatten(object v)
    {
        if (v is IDictionary<string, object?> d)
        {
            return new Dictionary<string, object?>(d);
        }
        throw new Refusal("the handler returned something that is not a set of reply fields");
    }
}
